### PR TITLE
Adds support for kubernetes_users, extend interpolation (#3404)

### LINKED ISSUE #3369: Add kube_users impersonation
### Kubernetes user

Sometimes there is a need to impersonate an existing IAM user. Existing teleport mapping only supports `kubernetes_groups`, however there is a need to support `kubernetes_user`.

Here is an example for IBM kubernets clusters:

```yaml
kind: role
spec:
  allow:
    kubernetes_user:
    - 'IAM#{{external.email}}'
```

This will result in `Impersonate-User` to be set to `IAM#sasha@ibm.com`

### PR DESCRIPTION
**Pull request by klizhentas**
This commit fixes #3369, refs #3374

It adds support for kuberenetes_users section in roles,
allowing Teleport proxy to impersonate user identities.

It also extends variable interpolation syntax by adding
suffix and prefix to variables and function `email.local`:

Example:

```yaml
kind: role
version: v3
metadata:
  name: admin
spec:
  allow:
    # extract email local part from the email claim
    logins: ['{{email.local(external.email)}}']

    # impersonate a kubernetes user with IAM prefix
    kubernetes_users: ['IAM#{{external.email}}']

  # the deny section uses the identical format as the 'allow' section.
  # the deny rules always override allow rules.
  deny: {}
```

Some notes on email.local behavior:

* This is the only function supported in the template variables for now
* In case if the email.local will encounter invalid email address,
it will interpolate to empty value, will be removed from resulting
output.

Changes in impersonation behavior:

* By default, if no kubernetes_users is set, which is a majority of cases,
  user will impersonate themselves, which is the backwards-compatible behavior.

* As long as at least one `kubernetes_users` is set, the forwarder will start
  limiting the list of users allowed by the client to impersonate.

* If the users' role set does not include actual user name, it will be rejected,
  otherwise there will be no way to exclude the user from the list).

* If the `kuberentes_users` role set includes only one user
  (quite frequently that's the real intent), teleport will default to it,
  otherwise it will refuse to select.

  This will enable the use case when `kubernetes_users` has just one field to
  link the user identity with the IAM role, for example `IAM#{{external.email}}`

* Previous versions of the forwarding proxy were denying all external
impersonation headers, this commit allows 'Impesrsonate-User' and
'Impersonate-Group' header values that are allowed by role set.

* Previous versions of the forwarding proxy ignored 'Deny' section of the roles
when applied to impersonation, this commit fixes that - roles with deny
kubernetes_users and kubernetes_groups section will not allow
impersonation of those users and groups.

### CODE REVIEW

**Comment by klizhentas** [comment]
@fspmarshall @russjones this is a forward port

**Review by russjones — APPROVED** [review]

File: lib/kube/proxy/forwarder.go [hunk 1]
```diff
@@ -652,16 +669,89 @@ func (f *Forwarder) setupForwardingHeaders(ctx *authContext, sess *clusterSessio
...
+	// By default, if no kubernetes_users is set (which will be a majority),
+	// user will impersonate themselves, which is the backwards-compatible behavior.
+	//
+	// As long as at least one `kubernetes_users` is set, the forwarder will start
+	// limiting the list of users allowed by the client to impersonate.
+	//
+	// If the users' role set does not include actual user name, it will be rejected,
+	// otherwise there will be no way to exclude the user from the list).
+	//
+	// If the `kuberentes_users` role set includes only one user
```
Comment (webvictim) [inline lib/kube/proxy/forwarder.go]: kubernetes

File: lib/services/types.proto [hunk 2]
```diff
@@ -566,6 +566,9 @@ message RoleConditions {
 	repeated string KubeGroups = 5 [(gogoproto.jsontag) = "kubernetes_groups,omitempty"];
 
     AccessRequestConditions Request = 6 [(gogoproto.jsontag) = "request,omitempty"];
+
+	// KubeUsers is an optional kuberentes users to impersonate
```
Comment (webvictim) [inline lib/services/types.proto]: kubernetes

**Review by webvictim — APPROVED** [review]

**Comment by klizhentas** [comment]
retest this please

**Pull request merged**